Feat/add internal playwright helpers

chelseapinka opened the pull request:
# Adding new package called @inrupt/internal-playwright-helpers
Includes reusable pages `CognitoPage`, `OpenIdPage` and `TestPage` and previously used utility functions such as `essUSerLogin` that has been renamed to `loginAndAllow` for removing duplicative code across our sdks.

chelseapinka committed:
Add existing models from the notifications pacakage

chelseapinka committed:
Update naming convention for exposed functions

chelseapinka committed:
Adding documentation about using package

chelseapinka committed:
Fix typo from copy and replace

NSeydoux commented:
Would it make sense, in addition to the playwright helpers, to export a React component exposing the buttons the helpers rely on? It was somewhat the intent of our shared styled component library, but it did too much for a test app. For instance, we could export something like https://github.com/inrupt/solid-client-notifications-js/blob/7103c3b7650d7b13fb17d361bfd4279b9bb2bd0f/e2e/browser/testApp/components/appContainer/index.tsx#L61, which I already copy-pasted from the notifications test to the access grants one in https://github.com/inrupt/solid-client-access-grants-js/pull/408

NSeydoux commented on packages/internal-playwright-helpers/README.md:
````diff
@@ -0,0 +1,17 @@
+Add the function import as needed.
+
+```
+import "@inrupt/internal-playwright-helpers"
````
```suggestion
import { loginAndAccept } from "@inrupt/internal-playwright-helpers"
```

NSeydoux commented:
Also, it may be out of scope for this PR, so feel free to add a ticket to the backlog for this, but `@inrupt/solid-client-authn` has a bunch of usefult fixtures to set up the tests: https://github.com/inrupt/solid-client-authn-js/blob/main/e2e/browser/fixtures.ts. That allows to use the Node API to create the resources we want to use in the tests and to clean them up afterwards, which would be a nice addition to these shared helpers.

NSeydoux commented on packages/internal-playwright-helpers/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "@inrupt/internal-playwright-helpers",
+  "version": "1.0.0",
+  "description": "This package provides page models for known common elements of the sdk testable UIs\"",
```
```suggestion
  "description": "This package provides page models for known common elements of the sdk testable UIs",
```

chelseapinka committed:
Update packages/internal-playwright-helpers/README.md

Co-authored-by: Zwifi <nseydoux@inrupt.com>

chelseapinka committed:
Update packages/internal-playwright-helpers/package.json

Co-authored-by: Zwifi <nseydoux@inrupt.com>

chelseapinka commented:
> Also, it may be out of scope for this PR, so feel free to add a ticket to the backlog for this, but `@inrupt/solid-client-authn` has a bunch of usefult fixtures to set up the tests: https://github.com/inrupt/solid-client-authn-js/blob/main/e2e/browser/fixtures.ts. That allows to use the Node API to create the resources we want to use in the tests and to clean them up afterwards, which would be a nice addition to these shared helpers.

I agree that would be useful. @ThisIsMissEm mentioned this as an option as well. I will go ahead and add a ticket for this but do we want it to be under the same JS e2e epic @NSeydoux?

chelseapinka committed:
Fix readme typo

NSeydoux commented:
We can add it to the epic, but not tag it for the next release, as it isn't something we strictly need, so we don't want to commit to it on the short term. One of these improve over time things.

NSeydoux commented on packages/internal-playwright-helpers/index.ts:
```diff
@@ -0,0 +1,77 @@
+
+  async startLogin() {
+    const { idp } = getBrowserTestingEnvironment();
+    await this.page.fill("[data-testid=identityProviderInput]", idp);
```
These helpers rely on these tag IDs, so it should probably be documented somewhere, or the associated components be exported (see comment https://github.com/inrupt/typescript-sdk-tools/pull/78#issuecomment-1324867490)

ThisIsMissEm replied:
@NSeydoux I just discussed this with @chelseapinka and we've a plan to tackle this:

- Add a `@inrupt/internal-playwright-testids` package with the testid strings
- Use those in both testApp's and `@inrupt/internal-playwright-helpers`

NSeydoux committed:
Add internal-playwright-testids package

This package exports testids that are shared between tests fixtures and test apps.

NSeydoux committed:
Use shared testids in playwright helpers

NSeydoux committed:
Setup automated login fixture

The fixture logs the user in by default, because that's what we want to do most of the time, but also add a flag to disable auto-login for tests requiring not logging in.

NSeydoux commented:
I added a fixture for automated login, see https://github.com/inrupt/solid-client-access-grants-js/pull/407/files for an example usage.

ThisIsMissEm commented on packages/internal-playwright-helpers/testApp.ts:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  private async startLogin() {
+    const { idp } = getBrowserTestingEnvironment();
```
We can pass this into the constructor for TestPage

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/78/commits/0e79133024ded1b6dccb0e8e7cda407111cefcd7

ThisIsMissEm commented on packages/internal-playwright-helpers/testApp.ts:
```diff
@@ -0,0 +1,95 @@
+      (request) =>
+        request.method() === "POST" && request.url().includes("/token")
+    );
+    await Promise.all([
```
Unnecessary Prkmise.all here

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/78/commits/da7ddc7d6f9621d3392dc8c27171ef1dce1cb0b7

ThisIsMissEm commented on packages/internal-playwright-helpers/tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "ES2017",
+    "module": "commonjs",
```
I think we can make this ESM only or SSNext

NSeydoux replied:
Does this address your suggestion: https://github.com/inrupt/typescript-sdk-tools/pull/78/commits/0f99267beb45c4720b6b2b81f52b424ccaebd465 ?

ThisIsMissEm replied:
I think so, yes.

NSeydoux committed:
Override the page fixture

Overriding the existing fixture rather than defining a new one allows to not change the existing tests

NSeydoux committed:
Fix rollup config

The previous config resulted in the `index.js` CJS file being overwritten by the `index.js` ESM one.

NSeydoux committed:
Update version to match monorepo

NSeydoux committed:
Pass the idp to the test app page model

NSeydoux committed:
Remove unused Promise.all

NSeydoux committed:
Do not rebuild test app page on login

NSeydoux committed:
Set module to ESNext

NSeydoux committed:
Fix some comments

NSeydoux committed:
Add some docs to the readme

NSeydoux committed:
Move orchestration to fixtures file

Keep each page focused on a single view, and move cross-page coordination to the fixtures file.

NSeydoux committed:
Reorganize files

Create a src/pages directory for pages, and move other files to src/

NSeydoux committed:
Refactor fixture

ThisIsMissEm commented on packages/internal-playwright-helpers/src/fixtures.ts:
```diff
@@ -0,0 +1,60 @@
+    await openIdPage.allow();
+    await testPage.handleRedirect();
+  }
+};
```
A better split might be to do:

```
const testPage = new TestPage(page, openidProvider);
await testPage.start()
```

In the `page` method of the Fixture, with just the login on start being whether you call `loginAndAllow` on TestPage or not.

ThisIsMissEm replied:
So the lines 28-33 would be inside a method on `TestPage` (loginAndAllow), and there's essentially no need for this method

NSeydoux replied:
That's how I initially approached, but it felt cleaner to have the test page be a page model similar to the others, without any extra dependencies, and then to have the multi-page coordination done at the fixture level. This would mean that adding new providers such as KeyCloak would require no change in the test page, and only in the fixtures, which seems like the right place.

ThisIsMissEm commented on packages/internal-playwright-helpers/src/pages/cognito.ts:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  async login(username: string, password: string) {
+    await this.page.fill(".visible-lg [type=text]", username);
```
We've previously not done it, but would it be worth verifying that `page.url()` is on the `amazoncognito.com` domain?

NSeydoux replied:
ea28581

ThisIsMissEm commented on packages/internal-playwright-helpers/src/pages/testApp.ts:
```diff
@@ -0,0 +1,76 @@
+      // It is important to call waitForNavigation before click to set up waiting.
+      this.page.waitForNavigation(),
+      // Clicking the link will indirectly cause a navigation.
+      this.page.click(`[data-testid=${TESTID_LOGIN_BUTTON}]`),
```
would it be worth the testids package exporting these selectors itself? e.g.,

```suggestion
      this.page.click(selectors.TESTID_LOGIN_BUTTON),
```

Instead

NSeydoux replied:
The package will need to export the raw testids in any case since they need to be imported by the test app, so I'm not sure also exporting the selectors would add much. So you think it would improve readability?

ThisIsMissEm replied:
Yeah, it's really abouf readability here & to ensure the testid isn't accidentally a prefix instead of the complete value

NSeydoux replied:
5b8f4f8

ThisIsMissEm commented on packages/internal-playwright-helpers/src/fixtures.ts:
```diff
@@ -0,0 +1,60 @@
+      clientCredentials.owner.login,
+      clientCredentials.owner.password
+    );
+    await use(page);
```
I think here we can actually do `await use(testPage)` such that when using this we can just do `page.loginAndAllow()` or `page.loginAndDeny()`

ThisIsMissEm replied:
That is, we're replacing the `page` fixture with our own augmented page fixture.

NSeydoux replied:
I'm not sure that will be an option: the `page` fixture signature expects an implementation of `Page` to be used, which `TestPage` isn't, it's a wrapper around it. If the tests require access to test-specific controls, then rather than overriding the fixture we should probably add a different one altogether typed as `TestApp`. However, if we do this, then we probably don't want to override the existing `page` fixture at all, but that means that the app will be started and logged in if and only f that new fixture is imported in the tests, e.g.
```
test("...", async ({page, testApp}) => {
  testApp.loginAndAllow();
});
```
However, that requires updating all the existing tests to include `testApp` in the destructured object, **even if it is'nt used in the test** so that the app automatically starts. That's what bit me when trying this out, and that's why I went towards overriding the `page` fixture, which was less awkward than having the unused object in the test call I think

NSeydoux replied:
As discussed in a follow-up conversation, we will approach this by having a separate `auth` fixture with high-level auth-related operations (e.g. complete login process), and having the base `page` fixture just automatically start the app (but not log in). This comes from not being able to extend the `Page` class, resulting in the `page` fixture not to expose additional methods.

NSeydoux commented on packages/internal-playwright-helpers/package.json:
```diff
@@ -0,0 +1,27 @@
+    }
+  },
+  "dependencies": {
+    "@inrupt/internal-playwright-testids": "^1.4.2",
+    "@inrupt/internal-test-env": "^1.4.2"
```
Oh interesting, but is this how it should be done even after these ship? I assumed Lerna was the one taking the wheel here.

ThisIsMissEm replied:
Off top of head, I'm honestly not sure. I just know elsewhere we're using file: links instead of fixed versions & that the publish resolves those to correct versions

NSeydoux committed:
Separate auth fixture from base page

Having both auth and startup in the same fixture leads to issues where it's hard to control for a given test whether it is authenticated or not. It is possible to control this behavior at the test suite granularity, but that's not really convenient. Also, the Playright typing system doesn't make it possible/easy to extend the Page class, which means the resulting fixture only has the base page properties, and that doesn't allow higher-level calls like "login" to go through all the login steps.

NSeydoux committed:
Expose ESM import in package.json

NSeydoux committed:
lint

NSeydoux committed:
Check cognito domain on login

NSeydoux committed:
Export selectors from testid package

chelseapinka commented on packages/internal-playwright-testids/package.json:
```diff
@@ -0,0 +1,26 @@
+  "license": "MIT",
+  "exports": {
+    ".": {
+      "require": "./dist/index.js",
```
:+1:

chelseapinka commented on packages/internal-playwright-helpers/src/pages/cognito.ts:
```diff
@@ -0,0 +1,53 @@
+export class CognitoPage {
+  page: Page;
+
+  static COGNITO_DOMAIN = "amazoncognito.com ";
```
My gut is telling me we shouldnt be hardcoding this, but its consistent across all envs so there is no real reason to right now.

NSeydoux replied:
Well it's not great, but since the whole concept of `pages` is pretty much hardcoding what we expect of the pages at least it kinda belongs there I guess

chelseapinka commented:
LGTM
